### PR TITLE
Add Speech Classification doc

### PR DESCRIPTION
**Pull request by fayejf**
Speech classification (Speech Command +VAD) docs

TODO: 
make long table pretty
move speech command and vad inference once asr transcribe instruction is ready to have consistency. 

Signed-off-by: fayejf <fayejf07@gmail.com>

### CODE REVIEW

File: docs/source/asr/sc/intro.rst [hunk 1]
```diff
@@ -0,0 +1,71 @@
+Speech Classification
+==================================
+Speech Classification refers to a set of tasks or problems of getting a program to automatically classify input utterance or audio segment into categories, 
+such as task Speech Command Recognition (multi-class), Voice Activity Detection (binary or multi-class), and Audio Sentiment Classification (typically multi-class), etc.
```
Comment (nithinraok) [inline docs/source/asr/sc/intro.rst:4]:
```suggestion
such as ~~task~~ Speech Command Recognition (multi-class), Voice Activity Detection (binary or multi-class), and Audio Sentiment Classification (typically multi-class), etc.
```

Comment (nithinraok) [inline docs/source/asr/sc/intro.rst:4]: remove task

File: docs/source/asr/sc/results.rst [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  nemo_asr.models.<MODEL_BASE_CLASS>.list_available_models()
+
+
+Speech Classification Models
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. tabularcolumns:: 30 30 40
+
+.. csv-table::
+   :file: data/classification_results.csv
```
Comment (nithinraok) [inline docs/source/asr/sc/results.rst:72]: Do we need speaker-related models here?

**Review by nithinraok — COMMENTED** [review]
two minor comments. Rest all looks good

File: docs/source/asr/sc/datasets.rst [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+If you have your own data and want to preprocess it to use with NeMo ASR models, check out the `Preparing Custom Speech Classification Data`_ section at the bottom of the page.
+
+.. _Freesound-dataset:
+
+Freesound
+-----------
+
+`Freesound <http://www.freesound.org/>`_ is a website that aims to create a huge open collaborative database of audio snippets, samples, recordings, bleeps. 
+Audio samples are released under Creative Commons licenses that allow their reuse.
```
Comment (titu1994) [inline docs/source/asr/sc/datasets.rst:15]: "Most audio samples ..."<- not all are CC-BY

File: docs/source/asr/sc/datasets.rst [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+`Freesound <http://www.freesound.org/>`_ is a website that aims to create a huge open collaborative database of audio snippets, samples, recordings, bleeps. 
+Audio samples are released under Creative Commons licenses that allow their reuse. 
+Researchers and developers can access Freesound content using the Freesound API to retrieve meaningful sound information such as metadata, analysis files, and the sounds themselves. 
+
+**Instructions**
+
+Go to ``<NeMo_git_root>/scripts/freesound_download_resample`` and follow the below steps to download and convert freedsound data into a format expected by the `nemo_asr` collection.
+
+1. We will need some requirements including freesound, requests, requests_oauthlib, joblib, librosa and sox. If they are not installed, please run `pip install -r freesound_requirements.txt`
```
Comment (titu1994) [inline docs/source/asr/sc/datasets.rst:22]: `some required libraries including ...`

File: docs/source/asr/sc/datasets.rst [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+Then, you should have 16khz mono wav files in `<resampled data directory>`. 
+
+
+.. _Google-Speech-Commands-Dataset:
+
+Google Speech Commands Dataset
+------------------------------
+
+Google released two versions of the dataset with the first version containing 65k samples over 30 classes and the second containing 110k samples over 35 classes.
+We refer to these datasets as v1 and v2
```
Comment (titu1994) [inline docs/source/asr/sc/datasets.rst:43]: `v1 and v2 respectively.`

File: docs/source/asr/sc/results.rst (hunk 2 above)
Comment (fayejf) [reply docs/source/asr/sc/results.rst:72]: Not really. good catch

File: docs/source/asr/sc/datasets.rst [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+Run the script `process_speech_commands_data.py` to process Google Speech Commands dataset in order to generate files in the supported format of  `nemo_asr`,
+which can be found in ``<NeMo_git_root>/scripts/dataset_processing/``. 
+You should set the data folder of Speech Commands using :code:`--data_root` and the version of the dataset using :code:`--data_version` as an int.
+
+You can further rebalance the train set by passing the `--rebalance` flag.
```
Comment (titu1994) [inline docs/source/asr/sc/datasets.rst:49]: `rebalance the train set by randomly oversampling files inside the manifest by passing the ...`

File: docs/source/asr/sc/datasets.rst [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+Each line is a training example.
+
+.. code-block:: bash
+
+  {"audio_filepath": "<absolute path to dataset>/two/8aa35b0c_nohash_0.wav", "duration": 1.0, "label": "two"}
+  {"audio_filepath": "<absolute path to dataset>/two/ec5ab5d5_nohash_2.wav", "duration": 1.0, "label": "two"}
+
+
+
+SCF dataset for VAD
```
Comment (titu1994) [inline docs/source/asr/sc/datasets.rst:72]: Full form of SCF (or at least note that SCF is a short form of the above dataset)

**Review by nithinraok — APPROVED** [review]
LGTM

File: docs/source/asr/sc/intro.rst [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+.. note::
+  For model details and deep understanding, please refer to tutorials in ``<NeMo_git_root>/tutorials/asr/``.
+
+
+Transcribe speech command segment:
+
+.. code-block:: python 
+
+  mbn_model = nemo_asr.models.EncDecClassificationModel.from_pretrained(<commandrecognition_en_matchboxnet3x1x64_v2>)
```
Comment (titu1994) [inline docs/source/asr/sc/intro.rst:29]: Make it generic like ASR - dont have explicit class name here. Remove explicit model name as well. Generics only.

File: docs/source/asr/sc/models.rst [hunk 9]
```diff
@@ -0,0 +1,71 @@
+Models
+======
+
+This page gives a brief overview of the models that NeMo's Speech Classification collection currently supports.
+Now for Speech Classification, we support Speech Command (Keyword) Detection and Voice Activity Detection (VAD).
```
Comment (titu1994) [inline docs/source/asr/sc/models.rst:5]: Drop the "Now"

File: docs/source/asr/sc/models.rst [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+For more information about the config files and how they should be structured, see the :doc:`./configs` page.
+
+Pretrained checkpoints for all of these models, as well as instructions on how to load them, can be found on the :doc:`./results` page.
+You can use the available checkpoints for immediate inference, or fine-tune them on your own datasets.
+The Checkpoints page also contains benchmark results for the available ASR models.
+
+.. _MatchboxNet-model:
+
+MatchboxNet (Commands)
```
Comment (titu1994) [inline docs/source/asr/sc/models.rst:19]: Speech Commands instead of Commands

File: docs/source/asr/sc/intro.rst [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+It serves to determine which samples to be sent to the model and when to close the microphone.
+
+
+``Quick Start``
+
+.. note::
+  For model details and deep understanding, please refer to tutorials in ``<NeMo_git_root>/tutorials/asr/``.
+
+
+Transcribe speech command segment:
```
Comment (titu1994) [inline docs/source/asr/sc/intro.rst:25]: Move transcription to results section - not in intro. Similar to ASR and Speaker

File: docs/source/asr/sc/models.rst [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+Similarly to QuartzNet, the MatchboxNet family of models are denoted as MatchBoxNet_[BxRxC] where B is the number of blocks, and R is the number of convolutional sub-blocks within a block, and C is the number of channels. Each sub-block contains a 1-D *separable* convolution, batch normalization, ReLU, and dropout:
+
+    .. image:: images/matchboxnet_vertical.png
+        :align: center
+        :alt: MatchboxNet model
+        :scale: 50%
+
+It can reach state-of-the art accuracy on the Google Speech Commands dataset while having significantly fewer parameters than similar models. 
+The _v1 and _v2 are denoted for models trained on v1 (30-way classification) and v2 (35-way classification) datasets;
```
Comment (titu1994) [inline docs/source/asr/sc/models.rst:32]: Add `` for `_v1` and `_v2`

File: docs/source/asr/sc/models.rst [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+Similarly to QuartzNet, the MatchboxNet family of models are denoted as MatchBoxNet_[BxRxC] where B is the number of blocks, and R is the number of convolutional sub-blocks within a block, and C is the number of channels. Each sub-block contains a 1-D *separable* convolution, batch normalization, ReLU, and dropout:
+
+    .. image:: images/matchboxnet_vertical.png
+        :align: center
+        :alt: MatchboxNet model
+        :scale: 50%
+
+It can reach state-of-the art accuracy on the Google Speech Commands dataset while having significantly fewer parameters than similar models. 
+The _v1 and _v2 are denoted for models trained on v1 (30-way classification) and v2 (35-way classification) datasets; 
+And we use _subset_task to represent (10+2)-way subset (10 specific classes + other remaining classes + silence) classification task.
```
Comment (titu1994) [inline docs/source/asr/sc/models.rst:33]: Same as above

File: docs/source/asr/sc/results.rst (hunk 2 above)
Comment (titu1994) [reply docs/source/asr/sc/results.rst:72]: +1

**Review by titu1994 — CHANGES_REQUESTED** [review]
Overall looks great. Minor fixes here and there. One issue is the path - why call it `asr/sr/*`. Please use `speech_classification` in full throughout. Similar to `speaker_recognition`.

File: docs/source/asr/sc/configs.rst [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+An example Speech Classification train and validation configuration could look like:
+
+.. code-block:: yaml
+
+  model:
+    sample_rate: 16000
+    repeat: 2
+    dropout: 0.0
+    kernel_size_factor: 1.0
+    labels: ['bed', 'bird', 'cat', 'dog', 'down', 'eight', 'five', 'four', 'go', 'happy', 'house', 'left', 'marvin',
```
[the remark this answers is not in the document]
Comment (fayejf) [reply docs/source/asr/sc/configs.rst:40]: I don't explicitly explain fine-tune (on subset) in doc (try to minimize) .  Let's keep the basic condition here. 
Will add a line in quick start to guide user to learn finetune  (on subset) in notebooks

File: docs/source/asr/sc/configs.rst [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+
+    train_ds:
+      manifest_filepath: ???
+      sample_rate: ${model.sample_rate}
+      labels: ${model.labels} # Uses the labels above
+      batch_size: 128
+      shuffle: True
+      is_tarred: False  # If set to true, uses the tarred version of the Dataset
+      tarred_audio_filepaths: null      # Not used if is_tarred is false
+      tarred_shard_strategy: "scatter"  # Not used if is_tarred is false
```
[the remark this answers is not in the document]
Comment (fayejf) [reply docs/source/asr/sc/configs.rst:52]: It does support. but yeah let's remove it here and point to asr config file. 
It's better to have a place clearly introduce the tarred dataset.

File: docs/source/asr/sc/intro.rst (hunk 11 above)
Comment (fayejf) [reply docs/source/asr/sc/intro.rst:25]: The speaker has embedding extraction here in intro and ASR transcribe part haven't finished (I think?)  Shall we just leave it here? If need to move, will move speaker reco, speech classification and speaker diarization together after I've seen ASR

File: docs/source/asr/datasets.rst [hunk 16]
```diff
@@ -156,6 +156,8 @@ Run the script to process AIShell-2 dataset in order to generate files in the su
 
 Then, you should have `train.json` `dev.json` `test.json` and `vocab.txt` in `dest_folder`. 
 
+
```
[the remark this answers is not in the document]
Comment (titu1994) [reply docs/source/asr/datasets.rst:159]: ++

File: docs/source/asr/speech_classification/configs.rst [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+An example Speech Classification train and validation configuration could look like: 
+
+.. code-block:: yaml
+
+  model:
+    sample_rate: 16000
+    repeat: 2 # number of convolutional sub-blocks within a block, R in <MODEL>_[BxRxC]
+    dropout: 0.0
+    kernel_size_factor: 1.0
+    labels: ['bed', 'bird', 'cat', 'dog', 'down', 'eight', 'five', 'four', 'go', 'happy', 'house', 'left', 'marvin',
```
Comment (titu1994) [inline docs/source/asr/speech_classification/configs.rst:39]: What about the 10+2 subset ?

File: docs/source/asr/speech_classification/intro.rst [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+``Quick Start``
+
+.. note::
+  For model details and deep understanding, please refer to tutorials in ``<NeMo_git_root>/tutorials/asr/``.
+
+
+Transcribe speech command segment:
+
+.. code-block:: python 
+
```
Comment (titu1994) [inline docs/source/asr/speech_classification/intro.rst:28]: Not in Intro - put this in transcribe into results section.

Comment (titu1994) [reply docs/source/asr/speech_classification/intro.rst:28]: If you really want a quick start - point to the notebooks here.

File: docs/source/asr/speech_classification/results.rst [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+In general, you can load any of these models with code in the following format.
+
+.. code-block:: python
+
+  import nemo.collections.asr as nemo_asr
+  model = nemo_asr.models.EncDecClassificationModel.from_pretrained(model_name="<MODEL_NAME>")
+
+Where the model name is the value under "Model Name" entry in the tables below.
+
+For example, to load the MatchboxNet3x2x64_v1 model for speech command detection, run:
```
Comment (titu1994) [inline docs/source/asr/speech_classification/results.rst:49]: No examples - ASR and rest dont have so much duplication. Revert this para

**Review by titu1994 — CHANGES_REQUESTED** [review]
Overall fine, still needs changes.

File: docs/source/asr/speech_classification/results.rst (hunk 19 above)
Comment (fayejf) [reply docs/source/asr/speech_classification/results.rst:49]: ASR has "For example, to load the base English QuartzNet model for speech recognition, run: ..." 
Do we need to delete it?

File: docs/source/asr/speaker_recognition/results.rst [hunk 20]
```diff
@@ -62,4 +62,23 @@ Speaker Recognition Models
    :file: data/speaker_results.csv
    :align: left
    :widths: 30, 30, 40
-   :header-rows: 1
\ No newline at end of file
+   :header-rows: 1
+
+
+
```
Comment (titu1994) [inline docs/source/asr/speaker_recognition/results.rst:68]: Inference should be above models table. IE model tables should be after all other documentation.

File: docs/source/asr/speech_classification/results.rst (hunk 19 above)
Comment (titu1994) [reply docs/source/asr/speech_classification/results.rst:49]: Hmm seems fine to keep then.

**Review by titu1994 — CHANGES_REQUESTED** [review]
Overall near perfect. Just push the docs *above* the tables in speaker and speech classification and were good to go.

**Review by titu1994 — APPROVED** [review]
Looks great ! Lets merge

**Pull request merged**